src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,546 @@
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getSnapshotTotalVotePower(
+    bytes32 hash
+  ) external view returns (uint256 totalVotePower) {
+    return _info[hash]._totalVotePow;
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getVote(bytes32 hash, address cid) external view returns (Vote v) {
+    return _info[hash]._vote[cid];
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getPendingProposals() external view returns (bytes32[] memory pendingHashes) {
+    bytes32[] memory allHashes = _pendingProposals.values();
+    uint256 length = allHashes.length;
+
+    uint256 count;
+    pendingHashes = new bytes32[](length);
+
+    for (uint256 i; i < length; ++i) {
+      if (getState(allHashes[i]) == State.Active) {
+        pendingHashes[count] = allHashes[i];
+        ++count;
+      }
+    }
+
+    assembly ("memory-safe") {
+      mstore(pendingHashes, count)
+    }
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getSnapshotVotePowerById(bytes32 hash, address cid) external view returns (uint256 votePower) {
+    return _info[hash]._votePower[cid];
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getMinimumVotePowerToPass(
+    bytes32 hash
+  ) external view returns (uint256 votePower) {
+    return _calcMinVoteFor(_info[hash]._totalVotePow);
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getMinimumVotePowerToCancel(
+    bytes32 hash
+  ) external view returns (uint256 minVoteAgainst) {
+    return _calcMinVoteAgainst(_info[hash]._totalVotePow);
+  }
+
+  /**
+   * @dev Create a new proposal.
+   */
+  function _createProposal(bytes32 hash, Proposal memory p, uint256 totalVotePower) internal {
+    ProposalInfo storage $ = _info[hash];
+
+    $._createdAt = uint40(block.timestamp);
+    $._state = State.Active;
+    $._proposal = p;
+    $._totalVotePow = (totalVotePower).toUint96();
+
+    require(_pendingProposals.add(hash), ErrExistedProposal(hash));
+
+    emit ProposalCreated(msg.sender, p.proposer, hash, p, $._createdAt);
+  }
+
+  /**
+   * @dev Snapshot the vote power of all candidates at the time of proposal creation.
+   */
+  function _snapshotVotePower(
+    bytes32 hash
+  ) internal returns (uint256 totalVotePower) {
+    address[] memory allCids = ICandidateManager(getContract(ContractType.VALIDATOR)).getValidatorCandidateIds();
+    uint256[] memory selfStakes = IStaking(getContract(ContractType.STAKING)).getManySelfStakingsById(allCids);
+
+    uint256 length = allCids.length;
+    require(length == selfStakes.length, ErrLengthMismatch(msg.sig));
+
+    ProposalInfo storage $ = _info[hash];
+
+    for (uint256 i; i < length; ++i) {
+      uint96 scaledStake = (selfStakes[i] / 1 ether).toUint96();
+      totalVotePower += scaledStake;
+
+      $._votePower[allCids[i]] = scaledStake;
+
+      // Self-stake is expected to > 1e18
+      require(scaledStake != 0, ErrPanicDataCorruption(allCids[i]));
```

Comment:
wait, if there's a new validator whose self staked is < 1e18, then it could block other candidate from proposing a proposal

Reply:
Then I expect revert here, since we have indicate there are bugs in our system